### PR TITLE
Update _authored_hints.haml

### PR DESCRIPTION
**Pull request by dancodedotorg**
Authored hints has this really wacky bug where XML is regenerated over and over again:

wild-bug

Which isn't great in general, and especially isn't great because this section is the only place where curriculum writers can test Feedback Strings before they go into production and make sure they don't look bad, which is [important to avoid discovering bad feedback strings after they've made their way to production](https://codedotorg.slack.com/archives/C04MKTHMBM1/p1682607882829039).

## I Fixed It! I think

Here's a video of how I diagnosed the bug and (hopefully) fixed it. I'd love it someone else could test it out in their environment too since I don't have a full dev setup.

https://user-images.githubusercontent.com/68714964/235571232-2bf8f94b-ab40-4766-ba86-0f32dafcb281.mp4

Feedback welcomed. Or, if this doesn't actually fix it, maybe it gets us 99.9% of the way there and someone can help me take it over the finish line!

## Warning!!

The AP CSP Create Performance Task is in progress from April 1 - May 1 2023. Please
consider any risk introduced by this PR that could affect our students taking AP CSP. Code.org
students taking AP CSP primarily use App Lab for their Create Task, however a small percent use Game
Lab. Carefully consider whether your change has any risk of alterering, changing, or breaking
anything in these two labs. Even small changes, such as a different button color, are considered
significant during this time period. Reach out to the Student Learning team or Curriculum team for
more details.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/views/levels/editors/fields/_authored_hints.haml [hunk 1]
```diff
@@ -67,6 +67,15 @@
             'markdown',
             {
               callback: function(editor, change) {
+                //Looking for `.markdown_preview div --> div --> [XML Blockly divs]` structure
+                //gets .markdown_preview div
```
Comment (breville) [inline dashboard/app/views/levels/editors/fields/_authored_hints.haml:71]: nit: comments should probabl y lo

File: dashboard/app/views/levels/editors/fields/_authored_hints.haml [hunk 2]
```diff
@@ -67,6 +67,15 @@
             'markdown',
             {
               callback: function(editor, change) {
+                //Looking for `.markdown_preview div --> div --> [XML Blockly divs]` structure
```
Comment (breville) [inline dashboard/app/views/levels/editors/fields/_authored_hints.haml]: nit: Recommend some slight styling changes.

From this:
```
                //Looking for `.markdown_preview div --> div --> [XML Blockly divs]` structure
                //gets .markdown_preview div
                let parent = space.find('.markdown_preview').get(0); 
```
to this:
```
                // Looking for `.markdown_preview div --> div --> [XML Blockly divs]` structure
                // gets .markdown_preview div.
                let parent = space.find('.markdown_preview').get(0); 

```

(with a blank line before the next statement)

File: dashboard/app/views/levels/editors/fields/_authored_hints.haml [hunk 3]
```diff
@@ -67,6 +67,15 @@
...
+                //gets .markdown_preview div
+                let parent = space.find('.markdown_preview').get(0); 
+                //When XML is inserted, it creates a smaller div
+                let subParent = parent.querySelector("div");
```
Comment (breville) [inline dashboard/app/views/levels/editors/fields/_authored_hints.haml]: Looking [here](https://github.com/code-dot-org/code-dot-org/blob/bc7ec7398e27a7bff334222ebb26613a9a086358/apps/src/templates/instructions/utils.js#L132-L137), it appears that it will be a `span` if the XML had a `<p>` for its parent.  Is that a possible scenario?

File: dashboard/app/views/levels/editors/fields/_authored_hints.haml [hunk 4]
```diff
@@ -67,6 +67,15 @@
...
+                //And then inside that div is where all of the Blockly XML divs are located that I want to delete
+                //If there aren't any divs: the querySelectors will gracefull fail without erroring because there's nothing to query or remove
+                subParent.querySelectorAll("div").forEach((e) => {e.remove()});
```
Comment (breville) [inline dashboard/app/views/levels/editors/fields/_authored_hints.haml]: Is it possible that this could remove `<div>` tags that are unrelated to the XML, and that we don't want to delete?

File: dashboard/app/views/levels/editors/fields/_authored_hints.haml (hunk 1 above)
Comment (dancodedotorg) [inline dashboard/app/views/levels/editors/fields/_authored_hints.haml]: @breville I might be missing something, but this comment this looks incomplete to me? Something got cut off?

Comment (breville) [reply dashboard/app/views/levels/editors/fields/_authored_hints.haml]: Oops, not sure what happened.  Anyway, was suggesting a space after each `//`, a `.` at the end of each sentence, and a blank line after the relevant code before the next comment.

Comment (breville) [reply dashboard/app/views/levels/editors/fields/_authored_hints.haml]: Oh, it's the same comment below, so I'll just remove this thread.

File: dashboard/app/views/levels/editors/fields/_authored_hints.haml (hunk 3 above)
Comment (mikeharv) [reply dashboard/app/views/levels/editors/fields/_authored_hints.haml]: Good find, Brendan, and yes it's possible. Testing an "inline" block on Dan's branch reveals the bug is still present because the workspace is in a span element:

https://user-images.githubusercontent.com/43474485/236836879-ca3e21cb-c26c-499d-a755-428c7d59ff04.mp4

File: dashboard/app/views/levels/editors/fields/_authored_hints.haml (hunk 4 above)
Comment (mikeharv) [reply dashboard/app/views/levels/editors/fields/_authored_hints.haml]: Yes, this is also possible. I'm not sure how often `<div>` tags are included in authored hints but they are supported today:
image
This change would result in removing those elements from the markdown preview:
image

**Comment by mikeharv** [comment]
@dancodedotorg So impressed that you were able to find the right place in our code base to fix this long-standing issue! Unfortunately, as @breville pointed out, we might also put blocks in a `<span>` element within a `<p>` if the blocks are meant to be rendered inline with other text. It can also get a little messy if we start removing `<div>` elements based on imprecise logic.

I think what we actually want to do here is remove _any_ element within the markdown preview tag if it includes a block SVG. I've pushed a [commit](https://github.com/code-dot-org/code-dot-org/pull/51638/commits/4a3d283965277288610a6d5f8d81388afd0e62af) to this branch that instead selects any `<svg>` elements that have both blocklySvg and readOnlyBlockSpace classes (which are observable from the browser console's Elements tab).

Here's a screenshot of the updated version working. It continues to support arbitrary div elements while preventing duplicate blocks of either type:
image

When the markdown is next changed, we will remove the `<span>` tag shown below, as well as the bottom-most `<div>`, but no others. Again, this selection is based on svg elements with the specific classes that should indicate something is a block.
image

**Review by breville — APPROVED** [review]
This latest iteration looks like a good improvement.

It does feel like there might be a more robust solution that moves the appropriate changes into `convertXmlToBlockly` (and makes that function aware that it might be working with a container that it's already seen before) but given the widespread use of that function, understood that it might be a risky change. 

It also feels like there is probably a truly next-generation solution that involves using React, but that's obviously beyond the scope of this one fix.
